### PR TITLE
API Routes and Endpoint Functions

### PR DESCRIPTION
**Pull request by julian-george**
Making basic, flexible API routes and endpoint functions for the models that are likely to need them.

### CODE REVIEW

**Review by ziruihao — APPROVED** [review]
Clean, organized, well-done! Thanks for building all of these!

File: packages/server/src/routes/forest-routes.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+const forestRouter = express.Router();
+
+forestRouter.post<{}, any, Forest>("/", async (req, res) => {
+  try {
+    await createForest(req.body);
+    res.status(201).send("Forest created.");
+  } catch (e: any) {
+    console.error(e);
+    res.status(500).send(e?.message ?? "Unknown error.");
```
Comment (ziruihao) [inline packages/server/src/routes/forest-routes.ts:13]: Good use of coalescing!

**Comment by ziruihao** [comment]
> Couple of general questions/points:
> 
> * With the params passed to the getter functions in the routes, should I use optional chaining? Here's an example of a place it may be necessary:
> 
> image
> 
> - I feel a lot of this could be DRYer, since I ended up copy-pasting a lot of this stuff to make different routes/services. I don't think it's a problem but maybe we should consider making some more general code if we think we'll need to add more models/routes - A bunch of models have non-pk "name" attributes, which I included in the respective params interface. However, I don't really see a use case where a user would be querying for, for example, a specific trip by its name, so maybe we could just remove that as a possible param.

`req.query` should always be defined, at the very least an empty object `{}`. So `req.query.forestId` is at worst case, `undefined`, which is fine to pass into the getter functions!

I agree there's a good amount of duplication in the routing part. We can do some abstraction and DRYing up by writing generic `req.query` parsers and getter functions. But that takes a bit of work and we'll have to wrestle with various edge cases. We probably would if this was for a bigger corporate project.

**Comment by julian-george** [comment]
> > Couple of general questions/points:
> > 
> > * With the params passed to the getter functions in the routes, should I use optional chaining? Here's an example of a place it may be necessary:
> > 
> > image
> > 
> > * I feel a lot of this could be DRYer, since I ended up copy-pasting a lot of this stuff to make different routes/services. I don't think it's a problem but maybe we should consider making some more general code if we think we'll need to add more models/routes - A bunch of models have non-pk "name" attributes, which I included in the respective params interface. However, I don't really see a use case where a user would be querying for, for example, a specific trip by its name, so maybe we could just remove that as a possible param.
> 
> `req.query` should always be defined, at the very least an empty object `{}`. So `req.query.forestId` is at worst case, `undefined`, which is fine to pass into the getter functions!
> 
> I agree there's a good amount of duplication in the routing part. We can do some abstraction and DRYing up by writing generic `req.query` parsers and getter functions. But that takes a bit of work and we'll have to wrestle with various edge cases. We probably would if this was for a bigger corporate project.

Yeah I agree with your last point for sure. Doing that extra abstraction would def be worth it if we will keep wrestling extensively with the backend but for now it looks like these will be most if not all the routes in the final product.

**Pull request merged**